### PR TITLE
BUILD: move all test dependencies to ./test_requirements.txt

### PR DESCRIPTION
**Pull request by mattip**
Move all the test dependencies to a single `test_requirements.txt`. Perhaps we should have a  `build_requirements.txt` with only cython?

The real goal is to enable testing whether these packages have updates, so I added use of `pip list -o` to list outdated packages.

Probably `pip install -r test_requirements.txt` should get a shoutout somewhere in the development documentation

### CODE REVIEW

File: tools/ci/appveyor/requirements.txt [hunk 1]
```diff
@@ -1,6 +0,0 @@
-cython
-nose
-pytest-timeout
-pytest-xdist
-pytest-env
-pytest-faulthandler
```
[the remark this answers is not in the document]
Comment (mattip) [reply tools/ci/appveyor/requirements.txt]: pytest-faulthandler is [part of core pytest](https://pypi.org/project/pytest-faulthandler/) starting from v5.0

File: tools/ci/appveyor/requirements.txt [hunk 2]
```diff
@@ -1,6 +0,0 @@
-cython
-nose
```
[the remark this answers is not in the document]
Comment (mattip) [reply tools/ci/appveyor/requirements.txt]: yes, and i would try to claim that we should avoid using it since it has been abandoned.

**Comment by mattip** [comment]
The Check_Dependencies test is correctly failing:
```
> Package    Version Latest Type 
> ---------- ------- ------ -----
> pytest     5.0.1   5.1.1  wheel
> pytest-cov 2.6.1   2.7.1  wheel
> pytz       2018.9  2019.2 wheel
```

Now we need to decide what to do.
- Leave it just like this. When any of the dependencies release an update, our CI will complain but we can ignore it until we update the dependencies in a separate PR
  - for: simple
  - against: makes it confusing for drive-by contributors when CI fails 
- Use a service instead of having the check in our CI. 
  - for: outsourcing the problem
  - against: github has no fine-grained permissions, any service will need complete control of the repo. Also, can we set it up to use this non-standard `test_requirements.txt`?
- Automate instead of fail the test: automatically issue a PR to update the dependencies
  - for: cleanly supports upgrading in a non-intrusive way
  - against: complicates the script, will need some API support from github which might break at some point

**Comment by mattip** [comment]
> Is it possible to mark that task as failable, so that we get an X from azure on just one CI integration, but that x does not aggregate to an overall GitHub ci failure?

I don't see how, the [Job spec](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/phases?view=azure-devops&tabs=yaml) does not have any "failable" option I can see

**Comment by mattip** [comment]
> I'd much prefer to use a service for this.

I am exploring options for sending an email when the test fails. I think we can add an extension for [SendMail](https://docs.microsoft.com/en-us/azure/sendgrid-dotnet-how-to-send-email), do we have an organizational email entity that can sign up with them to get a token?

**Comment by dopplershift** [comment]
Isn't issuing PRs to update dependencies what [Dependabot](https://dependabot.com/python/) is for?

**Comment by mattip** [comment]
It seems dependabot may work out. In order to get the ball rolling, can we merge this PR as is, so that we can set up dependabot (or another service) and point them to the new file?

**Comment by charris** [comment]
Merging so we can get started. This will break #14404, so I will update that. Thanks Matti.

**Pull request merged**